whitepaper/communication-protocol-between-blp-and-lp.puml

```diff
@@ -16,82 +21,146 @@ end box
 database "ledger" as ledger
 
 == Initialize connecting with a ledger ==
-rif -> rif: enumerate deployed Ledger Plugins
+admin -> rif: startService()
 activate rif
-rif -> verifier: initialization
+rif -> blp_registry: getLedgerPluginInfoList()
+activate blp_registry
+blp_registry --> rif: list of LedgerPluignInfo
+deactivate 
+rif -> verifier: connect()\n (with ValidatorInfo)
 activate verifier
-verifier -> validator: open socket.io connection
+verifier -> validator: open socket.io\n connection
 activate validator
-validator -> validator: check identity of verifier
-validator -> ledger: setup parameters to call a ledger node
+validator -> validator: authenticate Verifier
+validator -> ledger: setup parameters \nto call a ledger node
 activate ledger
 ledger --> validator: calling successed
 deactivate ledger
-validator --> verifier: connection established
+validator --> verifier: connection\n established
 deactivate validator
-verifier --> rif: initialization finished
+verifier --> rif: connect() OK
 deactivate verifier
+rif -> blp_registry: update LedgerPluginInfo\n as activated
+activate blp_registry
+blp_registry --> rif: update OK
+deactivate blp_registry
+rif -> verifier: startMonitor()
+activate verifier
+verifier -> verifier: start monitoring
+verifier --> rif: startMonitor() OK
+deactivate verifier
+rif --> admin: startService() OK
 deactivate rif
+...
 
 == Send a ledger operation request ==
 
-blp -> blp: event triggered
+euser -> rif: CREATE a trade
+activate rif
+rif -> rif: assign tradeID
```

Comment:
Would this mean that the routing interface has some sort of distributed storage associated with it so that it can manage the trade IDs?

Reply:
Yes. On my assumption, the routing interface needs some storage and it can manage the trade IDs.
Since tradeID is the only way to identify a businessLogicID, I'm sure that there should be a mechanism to properly route LedgerEvents that are uniquely assigned and notified externally and internally.
I think that this feature is hard to implement outside of the routing interface, is there an alternative plan?

Reply:
> Yes. On my assumption, the routing interface needs some storage and it can manage the trade IDs.
> Since tradeID is the only way to identify a businessLogicID, I'm sure that there should be a mechanism to properly route LedgerEvents that are uniquely assigned and notified externally and internally.

@takeutak Thank you for clarifying!

> I think that this feature is hard to implement outside of the routing interface, is there an alternative plan?

I agree it's hard to implement and I'm not against having it in the routing interface either, but do have a suggestion to somehow make it more generic which might be done just by simply naming the `tradeID` something more generic like `subscriptionId` or `operationId` or `correlationId` or `routingKey`, etc. the idea being that there can be use cases that do not involve actual trades but still need the same kind of functionality where observing clients need to hold onto a unique identifier for the routing purposes.
Do you think that renaming could make sense as I outlined above? (names are just ideas)

Do you have anything specific (code/technology stack/dependencies, etc) about the storage implementation for the routing interface already or is it just the design vision for now that's shown on the diagram? I don't want to hijack this thread with the implementation details because this is about the design, but would love to continue the discussion about the actual implementation as well.